fix: Custom Event entry is missing mandatory enum Event #232

ebukizy1 opened the pull request.

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,69 @@
+## What is Enum
+Enums, short for "enumerations," are a way to define a custom data type that consists of a fixed set of named values, called variants. Enums are useful for representing a collection of related values where each value is distinct and has a specific meaning. Enums are particularly useful in game development for representing game states, player actions, or any other set of related constants that a g ...
+```rust
+# [derive(Serde, Copy, Drop, Introspect, PartialEq, Print)]
````
Perhaps we may not want to have all those derive here, while only introducing the concept?

glihm replied:
I would rather add a comment or reference to some Cairo documentation about those attributes to not have them in a simple example, wdyt?

glihm replied:
`Introspect` is necessary if the enum must be into a model.
`Serde` is necessary if the enum is used as input/output of an external function of a dojo contract.
`Copy`, `Drop`, `PartialEq`, `Print` may vary depend on the use-case.

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,69 @@
 
-You can define an enum as follows:
+```
+In this example, we've defined an enum called PlayerCharacter with four variants: `Godzilla`, `Dragon`, `Fox`, and `Rhyno`. The naming convention is to use PascalCase for enum variants. Each variant represents a distinct value of the PlayerCharacter type. In this particular example, variants don't have any associated value. 
````
```suggestion
In this example, we've defined an enum called PlayerCharacter with four variants: `Godzilla`, `Dragon`, `Fox`, and `Rhyno`. The naming convention is to use `PascalCase` for enum variants. Each variant represents a distinct value of the `PlayerCharacter` type. In this particular example, variants don't have any associated value. 
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,69 @@
 
 ```rust
+# [derive(Serde, Copy, Drop, Introspect, PartialEq, Print)]
+enum PlayerCharacter {
+    Godzilla :u128,
+    Dragon : u128,
+    Fox : u128,
+    Rhyno : u128
+}
````
Inconsistent format. IIRC, we should have something like `Dragon: u128`.

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,69 @@
+
+```
+
+In the next example, we will see that it is also possible to associate different data types with each variant. The PlayerCharacter enum we will define serves as a blueprint for creating various types of characters, each with their unique attributes. This design choice is particularly useful in game development, where you often need to manage different types of entities (like characters, items, or ...
+
+``` rust 
+# [derive(Serde, Copy, Drop, Introspect, PartialEq, Print)]
+enum PlayerCharacter {
+    Godzilla ,
+    Fox  : felt252,
+}
+
+    // Instantiate some characters with associated values
+let godzilla = PlayerCharacter::Godzilla;
+let fox = PlayerCharacter::Fox(90);
+```
````
This is not supported in Dojo yet. All enum variants must have the same type for now.

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -1,21 +1,69 @@
+
+## Trait Implementations for Enums
+
+ traits are a way to define shared behavior across types. By defining traits and implementing them for your custom enums, you can encapsulate common behaviors and methods that are relevant to the enum. This approach enhances code reusability and maintainability, especially in complex systems like game development.
```
```suggestion
 Traits are a way to define shared behavior across types. By defining traits and implementing them for your custom enums, you can encapsulate common behaviors and methods that are relevant to the enum. This approach enhances code reusability and maintainability, especially in complex systems like game development.
```

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -1,21 +1,69 @@
+
+ traits are a way to define shared behavior across types. By defining traits and implementing them for your custom enums, you can encapsulate common behaviors and methods that are relevant to the enum. This approach enhances code reusability and maintainability, especially in complex systems like game development.
+
+Consider the GameStatus enum, which represents the various states a game can be in. This enum is a simple yet powerful example of how enums can be used to model game states.
```
```suggestion
Consider the `GameStatus` enum, which represents the various states a game can be in. This enum is a simple yet powerful example of how enums can be used to model game states.
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,69 @@
+```
 
-// We define an into trait
+<!-- We define an into trait -->
````
From where did you take the `<!-- -->`, html I guess? :)
In cairo, we use `//`. 👍

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -28,8 +76,8 @@ impl GameStatusFelt252 of Into<GameStatus, felt252> {
-Then within a trait you can create something like this:
-
+# Structuring Game Logic with Traits and enum
+Building upon the GameStatus enum, we can define a Game struct that includes a GameStatus field. By implementing a custom trait for the Game struct, we can encapsulate game-specific logic and assertions.
```
```suggestion
Building upon the `GameStatus` enum, we can define a Game struct that includes a `GameStatus` field. By implementing a custom trait for the Game struct, we can encapsulate game-specific logic and assertions.
```

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+## Enum Placement
+Regarding the placement of the enum, it's generally best practice to define enums in the same file or module where they are used, especially if they are closely tied to the functionality of that system. This makes the code easier to understand and maintain, as the enum is defined in context. However, if the enum is used across multiple systems or components, it might make sense to define it in a  ...
+
+Given that Eternum has one enum in each system, it suggests a design where each system is self-contained and has its own set of related events or states. This design can help to encapsulate the logic of each system, making the codebase easier to navigate and understand.
```
We may want to add a link to the actual Eternum code.

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+Given that Eternum has one enum in each system, it suggests a design where each system is self-contained and has its own set of related events or states. This design can help to encapsulate the logic of each system, making the codebase easier to navigate and understand.
+
+## Important Of Enums 
+1 Semantic Clarity:
```
```suggestion
1. Semantic Clarity:
```
I guess here without the `1.` the rendering will not consider it as a list.

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+## Important Of Enums 
+1 Semantic Clarity:
+Enums provide semantic clarity by giving meaningful names to specific values. Instead of using arbitrary integers or strings, you can use descriptive identifiers.
+For example, consider an enum representing different player states: `IDLE`, `RUNNING`, `JUMPING`, and `ATTACKING`. These names convey the purpose of each state more effectively than raw numeric values.
```
```suggestion
For example, consider an enum representing different player states: `Idle`, `Running`, `Jumping`, and `Attacking`. These names convey the purpose of each state more effectively than raw numeric values.
```
Consistent casing with the `PascalCase` for variants.

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+Enums provide semantic clarity by giving meaningful names to specific values. Instead of using arbitrary integers or strings, you can use descriptive identifiers.
+For example, consider an enum representing different player states: `IDLE`, `RUNNING`, `JUMPING`, and `ATTACKING`. These names convey the purpose of each state more effectively than raw numeric values.
+
+2 Avoiding Magic Numbers:
```
```suggestion
2. Avoiding Magic Numbers:
```

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+    EnemyDefeated,
+    ItemCollected,
+}
+Now, when handling events, you can use Event::PlayerSpawned instead of an arbitrary number.
```
```suggestion

// Now, when handling events, you can use Event::PlayerSpawned instead of an arbitrary number.
```
Or we may put it outside.

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+Now, when handling events, you can use Event::PlayerSpawned instead of an arbitrary number.
+```
+
+3 Type Safety:
````
```suggestion
3. Type Safety:
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+```
+
+3 Type Safety:
+Enums provide type safety. Each enum variant has a distinct type, preventing accidental mixing of incompatible values.
````
```suggestion
Enums provide type safety. Each enum variant has a type, preventing accidental mixing of incompatible values.
```

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+    SpeedBoost,
+    Invincibility,
+}
+You can’t mistakenly assign a PowerUp value to a variable expecting a different type.
```
```suggestion
// You can’t mistakenly assign a PowerUp value to a variable expecting a different type.
```
Or place it outside.

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+}
+You can’t mistakenly assign a PowerUp value to a variable expecting a different type.
+```
+4 Pattern Matching:
````
```suggestion
4. Pattern Matching:
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+```
+4 Pattern Matching:
+Enums shine when used in pattern matching (also known as switch/case statements).
+You can handle different cases based on the enum variant, making your code more expressive and concise. example
````
```suggestion
You can handle different cases based on the enum variant, making your code more expressive and concise. Example:
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+
+```
+
+5 Extensibility:
````
```suggestion
5. Extensibility:
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+Enums allow you to add new variants without breaking existing code.
+Suppose you later introduce a DoubleDamage power-up. You can simply extend the PowerUp enum:
+
+```Rust
````
```suggestion
```rust
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -52,4 +100,73 @@ impl GameImpl of GameTrait {
+}
+```
+
+ enums serve as powerful tools for creating expressive, self-documenting code. They enhance readability, prevent errors, and facilitate better software design.
````
```suggestion
 Enums serve as powerful tools for creating expressive, self-documenting code. They enhance readability, prevent errors, and facilitate better software design.
```

glihm reviewed (changes requested):
Awesome writing here @ebukizy1!
Appreciate your time on that. I've left some comments, please let me know if you have some questions. 👍

ebukizy1 commented:
Thank you for your guidance! Your feedback is straightforward and greatly appreciated. I'll implement the changes and submit a pull request

glihm commented:
> Thank you for your guidance! Your feedback is straightforward and greatly appreciated. I'll implement the changes and submit a pull request

Feel free to apply those changes to your current pull request, and pushing again on the same branch will update everything. Let me know if you have question to this regard. 👍

ebukizy1 commented:
I've added all the necessary changes and pushed the updates. Please take a moment to review them and let me know if any corrections are needed.

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -2,53 +2,60 @@
-    Dragon : u128,
-    Fox : u128,
-    Rhyno : u128
+    Godzilla(CharacterAttributes),
+    Fox(CharacterAttributes),
+    Rhyno(CharacterAttributes),
```
Super nice, elegant solution to showcase how this can be done. 👍

One the dojo storage layout will be re-work, we should be able to support natively enums with different types.

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,76 @@
+});
+```
+For the `Godzilla` variant, since it doesn't require any additional data, you would instantiate it with a `CharacterAttributes` struct, where the `is_godzilla` flag is set to true, and the other fields are set to None or their default values. This approach effectively addresses the limitation where enum variants must share the same type.
+the `PlayerCharacter` enum has three variants: `Godzilla`, `Fox`, and `Rhyno`. Each variant takes a `CharacterAttributes` struct as its associated value. The `CharacterAttributes` struct contains fields to represent the unique attributes of each character type:
````
```suggestion
The `PlayerCharacter` enum has three variants: `Godzilla`, `Fox`, and `Rhyno`. Each variant takes a `CharacterAttributes` struct as its associated value. The `CharacterAttributes` struct contains fields to represent the unique attributes of each character type:
```

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -1,21 +1,76 @@
-// This enum simply defines the states of a game.
-#[derive(Serde, Copy, Drop, Introspect, PartialEq, Print)]
+`is_godzilla`: A boolean flag to indicate if the character is Godzilla.
+`fox_value: An`  `optional u32` value to represent the `felt252` value for Fox.
```
```suggestion
`fox_value`: An optional `u32` value to represent the `felt252` value for Fox.
```

glihm commented on docs/pages/cairo/enum.md:
````diff
@@ -1,21 +1,76 @@
+    rhyno_values: None,
+});
+```
+For the `Godzilla` variant, since it doesn't require any additional data, you would instantiate it with a `CharacterAttributes` struct, where the `is_godzilla` flag is set to true, and the other fields are set to None or their default values. This approach effectively addresses the limitation where enum variants must share the same type.
````
```suggestion
For the `Godzilla` variant, since it doesn't require any additional data, you would instantiate it with a `CharacterAttributes` struct, where the `is_godzilla` flag is set to true, and the other fields are set to `None` or their default values. This approach effectively addresses the limitation where enum variants must share the same type.
```

glihm commented on docs/pages/cairo/enum.md:
```diff
@@ -52,4 +107,71 @@ impl GameImpl of GameTrait {
+
+ enums serve as powerful tools for creating expressive, self-documenting code. They enhance readability, prevent errors, and facilitate better software design.
+
 > Read more about Cairo enums [here](https://book.cairo-lang.org/ch06-00-enums-and-pattern-matching.html)
```
```suggestion
Read more about Cairo enums [here](https://book.cairo-lang.org/ch06-00-enums-and-pattern-matching.html)
```

glihm reviewed (commented):
Nice! The work around for current Dojo limitation is great!

Also, some of the previous review comments are still here, I let them as non-resolve. Let me know if you have additional questions!

ebukizy1 commented:
I've implemented all the changes and added the necessary link for the enum attribute. Thank you very much for explaining the purpose of those imported drives. Could you please review it and let me know if there are any further adjustments needed? I appreciate your guidance

ebukizy1 commented:
Good day sir.. regarding the update I made haven’t gotten any response to know if there is any minor changes I should do or it accepted..

ebukizy1 commented:
good day haven't gotten feed back since 4 days i made my correction just concern so i can know if it is merge or let me do some changes

glihm commented:
> good day haven't gotten feed back since 4 days i made my correction just concern so i can know if it is merge or let me do some changes

Gm! I went by the PR during the last days but I didn't see updates on some comments. So I was wondering if you were still correcting them.
Could you double check that all comments are correctly addressed? Or are some changes missing to be pushed?

(example of missing comments: https://github.com/dojoengine/book/pull/255/files#r1562825102)

ebukizy1 commented:
Good day! I've revisited the code and made final changes, including removing multiple data types from the enum. Please review it and let me know if any further adjustments are needed. I believe I've addressed all the changes
.

ebukizy1 commented:
"Sir, I'm not sure if the changes I made in the push have been reflected. I would greatly appreciate it if you could review them. This would help me understand if there are any additional modifications needed or if everything is complete. Additionally, I would like to know if the pull request (PR) accurately reflects the changes I've made."

ebukizy1 commented:
Good day sir.. chatting you regarding the pr it seems you didn’t see this one too?? Because did all the correction you pointed out so I don’t know if it did not reflect so please sir will love to get your response on if the correction is accepted or I will still do modification

glihm reviewed (approved):
@ebukizy1 thank you for the time here and sorry for the delay, @gianalarcon appreciate the latest changes. 👍
